// @locator/runtime/dist/functions/buildLink.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.buildLink = buildLink;
exports.buildLinkFromSource = buildLinkFromSource;
exports.getSavedProjectPath = getSavedProjectPath;
exports.setInternalProjectPath = setInternalProjectPath;

var _linkTemplateUrl = require("./linkTemplateUrl");

var _evalTemplate = require("./evalTemplate");

let internalProjectPath = null;

function setInternalProjectPath(projectPath) {
  internalProjectPath = projectPath;
}

function getSavedProjectPath(options) {
  return options.getOptions().projectPath || internalProjectPath;
}

function buildLink(linkProps, targets, options, localLinkTypeOrTemplate) {
  const params = {
    filePath: linkProps.filePath,
    projectPath: getSavedProjectPath(options) || linkProps.projectPath,
    line: String(linkProps.line),
    column: String(linkProps.column)
  };
  return (0, _evalTemplate.evalTemplate)((0, _linkTemplateUrl.linkTemplateUrl)(targets, options, localLinkTypeOrTemplate), params);
}

function buildLinkFromSource(source, targets, options) {
  return buildLink({
    filePath: source.fileName,
    projectPath: source.projectPath || "",
    line: source.lineNumber,
    column: source.columnNumber || 0
  }, targets, options);
}
