// @locator/runtime/dist/adapters/svelte/svelteAdapter.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.SvelteTreeNodeElement = void 0;
exports.getElementInfo = getElementInfo;

var _goUpByTheTree = require("../goUpByTheTree");

var _HtmlElementTreeNode = require("../HtmlElementTreeNode");

function getElementInfo(found) {
  if (found.__svelte_meta) {
    const {
      loc
    } = found.__svelte_meta;
    return {
      thisElement: {
        box: found.getBoundingClientRect(),
        label: found.nodeName.toLowerCase(),
        link: {
          column: loc.column + 1,
          line: loc.line + 1,
          filePath: loc.file,
          projectPath: ""
        }
      },
      htmlElement: found,
      parentElements: [],
      componentBox: found.getBoundingClientRect(),
      componentsLabels: []
    };
  }

  return null;
}

class SvelteTreeNodeElement extends _HtmlElementTreeNode.HtmlElementTreeNode {
  getSource() {
    const element = this.element;

    if (element.__svelte_meta) {
      const {
        loc
      } = element.__svelte_meta;
      return {
        fileName: loc.file,
        lineNumber: loc.line + 1,
        columnNumber: loc.column + 1
      };
    }

    return null;
  }

  getComponent() {
    return null;
  }

}

exports.SvelteTreeNodeElement = SvelteTreeNodeElement;

function getTree(element) {
  const originalRoot = new SvelteTreeNodeElement(element);
  return (0, _goUpByTheTree.goUpByTheTree)(originalRoot);
}

const svelteAdapter = {
  getElementInfo,
  getTree
};
var _default = svelteAdapter;
exports.default = _default;
