// @locator/runtime/dist/components/IntroInfo.cjs
"use strict";

var _$template = require("solid-js/web").template;

var _$delegateEvents = require("solid-js/web").delegateEvents;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.IntroInfo = IntroInfo;

var _web = require("solid-js/web");

var _shared = require("@locator/shared");

var _solidJs = require("solid-js");

var _bannerClasses = require("../functions/bannerClasses");

var _BannerHeader = _interopRequireDefault(require("./BannerHeader"));

var _isCombinationModifiersPressed = require("../functions/isCombinationModifiersPressed");

var _optionsStore = require("../functions/optionsStore");

const _tmpl$ = /*#__PURE__*/_$template(`<div><div class="text-sm mt-2 mb-1">Go to component code with <!> + <div class="inline-block px-1 py-0.5 border border-slate-200 rounded">click</div> </div><div class="text-xs mt-2 mb-1 text-gray-600 flex gap-1"><a class="underline cursor-pointer" href="https://www.locatorjs.com" target="_blank">What is Locator?</a><a class="underline cursor-pointer">Stop showing this popup</a></div></div>`, 13),
      _tmpl$2 = /*#__PURE__*/_$template(`<div class="inline-block px-1 py-0.5 border border-slate-200 rounded"></div>`, 2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function IntroInfo(props) {
  const options = (0, _optionsStore.useOptions)();
  const [showIntro, setShowIntro] = (0, _solidJs.createSignal)(true);
  setTimeout(() => {
    setShowIntro(false);
  }, 5000);
  (0, _solidJs.createEffect)(() => {
    if (props.hide && showIntro()) {
      setShowIntro(false);
    }
  });

  const modifiers = () => (0, _isCombinationModifiersPressed.getMouseModifiers)();

  return (() => {
    const _el$ = _tmpl$.cloneNode(true),
          _el$2 = _el$.firstChild,
          _el$3 = _el$2.firstChild,
          _el$6 = _el$3.nextSibling,
          _el$5 = _el$6.nextSibling,
          _el$7 = _el$2.nextSibling,
          _el$8 = _el$7.firstChild,
          _el$9 = _el$8.nextSibling;

    (0, _web.insert)(_el$, (0, _web.createComponent)(_BannerHeader.default, {
      get openOptions() {
        return props.openOptions;
      },

      get adapter() {
        return props.adapter;
      }

    }), _el$2);
    (0, _web.insert)(_el$2, (0, _web.createComponent)(_solidJs.For, {
      get each() {
        return Object.keys(modifiers());
      },

      children: (key, i) => {
        return [(0, _web.memo)(() => i() === 0 ? "" : " + "), (() => {
          const _el$10 = _tmpl$2.cloneNode(true);

          (0, _web.insert)(_el$10, () => _shared.modifiersTitles[key]);
          return _el$10;
        })()];
      }
    }), _el$6);

    _el$9.$$click = () => {
      options.setOptions({
        showIntro: false
      });
    };

    (0, _web.effect)(_p$ => {
      const _v$ = (0, _bannerClasses.bannerClasses)(),
            _v$2 = showIntro() ? "12px" : "-120px";

      _v$ !== _p$._v$ && (0, _web.className)(_el$, _p$._v$ = _v$);
      _v$2 !== _p$._v$2 && _el$.style.setProperty("bottom", _p$._v$2 = _v$2);
      return _p$;
    }, {
      _v$: undefined,
      _v$2: undefined
    });
    return _el$;
  })();
}

_$delegateEvents(["click"]);
