// @locator/runtime/dist/adapters/react/fiberToSimple.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.fiberToSimple = fiberToSimple;

var _getBoundingRect = require("../../functions/getBoundingRect");

var _getComposedBoundingBox = require("../../functions/getComposedBoundingBox");

var _getUsableName = require("../../functions/getUsableName");

var _getAllFiberChildren = require("./getAllFiberChildren");

var _makeFiberId = require("./makeFiberId");

function fiberToSimple(fiber, manualChildren) {
  let simpleChildren;

  if (fiber.elementType?.styledComponentId) {
    const children = (0, _getAllFiberChildren.getAllFiberChildren)(fiber);

    if (children.length === 1) {
      const simple = fiberToSimple(children[0], manualChildren);
      simple.name = `${simple.name} (styled)`;
      return simple;
    }
  }

  if (manualChildren) {
    simpleChildren = manualChildren;
  } else {
    const children = (0, _getAllFiberChildren.getAllFiberChildren)(fiber);
    simpleChildren = children.map(child => {
      return fiberToSimple(child);
    });
  }

  const element = fiber.stateNode instanceof Element || fiber.stateNode instanceof Text ? fiber.stateNode : fiber.stateNode?.containerInfo;

  if (element) {
    const box = (0, _getBoundingRect.getBoundingRect)(element);
    return {
      type: "element",
      element: element,
      fiber: fiber,
      uniqueId: (0, _makeFiberId.makeFiberId)(fiber),
      name: (0, _getUsableName.getUsableName)(fiber),
      box: box || (0, _getComposedBoundingBox.getComposedBoundingBox)(simpleChildren),
      children: simpleChildren,
      source: fiber._debugSource || null
    };
  } else {
    return {
      type: "component",
      fiber: fiber,
      uniqueId: (0, _makeFiberId.makeFiberId)(fiber),
      name: (0, _getUsableName.getUsableName)(fiber),
      box: (0, _getComposedBoundingBox.getComposedBoundingBox)(simpleChildren),
      children: simpleChildren,
      source: fiber._debugSource || null,
      definitionSourceFile: simpleChildren.reduce((acc, curr) => {
        if (curr.source?.fileName) {
          return curr.source?.fileName;
        } else {
          return acc;
        }
      }, null)
    };
  }
}
