// @locator/runtime/dist/adapters/HtmlElementTreeNode.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HtmlElementTreeNode = void 0;

var _getReferenceId = require("../functions/getReferenceId");

var _nonNullable = _interopRequireDefault(require("../functions/nonNullable"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable @typescript-eslint/ban-ts-comment */
class HtmlElementTreeNode {
  type = "element";

  constructor(element) {
    this.element = element;
    this.name = element.nodeName.toLowerCase();
    this.uniqueId = String((0, _getReferenceId.getReferenceId)(element));
  }

  getBox() {
    return this.element.getBoundingClientRect();
  }

  getElement() {
    return this.element;
  }

  getChildren() {
    const children = Array.from(this.element.children);
    return children.map(child => {
      if (child instanceof HTMLElement) {
        // @ts-ignore
        return new this.constructor(child);
      } else {
        return null;
      }
    }).filter(_nonNullable.default);
  }

  getParent() {
    if (this.element.parentElement) {
      // @ts-ignore
      return new this.constructor(this.element.parentElement);
    } else {
      return null;
    }
  }

  getSource() {
    throw new Error("Method not implemented.");
  }

  getComponent() {
    throw new Error("Method not implemented.");
  }

}

exports.HtmlElementTreeNode = HtmlElementTreeNode;
